README.md
# teletext-twitter
Display Twitter timelines or searches for tweets and turns them into teletext pages for your Raspberry Pi

![Screenshot](https://i.imgur.com/yPKagm8.jpg "Screenshot 1")

## What is this?
Useless, for a start.. But if you've ever had the desire to read your tweets by pushing the teletext button on your remote and keying a page number then this is the package for you! Retro chic for the social media generation!

## Installation
After cloning the repository you'll need to install the Python-Twitter wrapper with pip:
`pip install python-twitter`

Before you get this up and running there are two other things you need to do first:

1) Connect your Raspberry Pi to a TV over composite using a cable with the correct pin-out. I bought mine from The Pi Hut: https://thepihut.com/products/adafruit-a-v-and-rca-composite-video-audio-cable-for-raspberry-pi
2) Install the VBIT2 and raspi-teletext apps and make sure they are outputting teletext data to your TV

* raspi-teletext (Alistair Buxton): https://github.com/ali1234/raspi-teletext - Only PAL is supported by raspi-teletext
* VBIT2 (Peter Kwan): https://github.com/peterkvt80/vbit2

## Setup
After getting everything installed there's one more setup task to do. Rename config.py-default to config.py and open the file in a text editor.

Now add your Twitter access tokens to the top section. You can find a good guide for doing this here - https://iag.me/socialmedia/how-to-create-a-twitter-app-in-8-easy-steps/ You will need to pick a unique name for the app. Which is annoying. Pick anything you want that isn't taken. Maybe add your name at the end.

## Usage

When you've setup your config.py you can run the script like this example that grabs your home timeline:

`python3 teletext-twitter -m home` (add -h to show options - also listed below)

The script will constantly update your chosen page number (default is 153 - chosen because it used to be used for this purpose on Teefax in the past) in the main teletext folder (which defaults in VBIT2 to /home/pi/teletext/).

All of the files in that folder are sent across to the TV every so often, therefore the script constantly overwrites it with new tweets so that it will update on your screen. Up to 99 subpages of tweets will be displayed - any further tweets will be dropped. Tweets are truncated for space reasons. You kinda have to, really...

## Config options

As well as your Twitter access tokens the following things can be changed in be config.py file:

### Theme support
In the config.py file you can see a theme section. The teletext level I am using supports 7 colours:
* Red 
* Green
* Yellow
* Blue
* Magenta
* Cyan
* White

You can change the colours of:
- the main header bar
- the thin separator bar
- the page title
- tweeter's usernames
- tweet timestamps
- the text of the tweet itself
- the highlight used on search queries

Finally, the title text string and colour at the top can be changed, although there is a 30 character limit or things look wonky.

### Miscellaneous options

- tti_path: (default: "/home/pi/teletext/") - the path your pages will be saved to
- page_number: (default: 153) - the page number to use
- cycle_time: (default: 20) - approximate time in seconds between subpages

## Notes
* New tweets are grabbed every 60 seconds by default. This is configurable with the -d option, but you do have be aware of the Twitter API limits. 60 seconds is the minimum
* Up to 99 subpages are written.
* Emoji stripping is now included. Mostly.
* URLs are stripped and replaced with the placeholder [LINK]
* Level 1.5 teletext is supported - meaning more characters are now supported! There you go, Europeans!

Apart from those notes, things should work ok. Have fun, turn back the clock, and if you genuinely use this for anything please let me know (also you're mental/cool).

### -h/--help output

<pre>
usage: teletext-twitter [-h] [-m MODE] [-q QUERY] [-c COUNT] [-d DELAY] [-n]
                        [-v] [-Q]

Reads your Twitter timeline and turns it into teletext pages for your
Raspberry Pi.

optional arguments:
  -h, --help            show this help message and exit
  -m MODE, --mode MODE  choose between different modes - home, user or search
  -q QUERY, --query QUERY
                        a search query, either a search term or a username.
                        hashtags supported if you put quotes around the string
  -c COUNT, --count COUNT
                        number of tweets to download (default is 5, capped at
                        200)
  -d DELAY, --delay DELAY
                        seconds between timeline scrapes (default is 60
                        seconds - lower values have no effect)
  -n, --no-repeat       only download tweets once - overrules -d switch
  -v, --version         show program's version number and exit
  -Q, --quiet           suppresses all output to the terminal except warnings
                        and errors
</pre>


changelog.md
## Changelog

## v1.1.0
- a major update! alistair cree had added level 1.5 teletext support
  this effectively means that more characters are supported. Europe rejoice!

### v1.0.4
- added query highlighting in search mode

### v1.0.3
- script now exits with correct returncode in no-repeat mode

### v1.0.2
- removed debug message

### v1.0.1
- added a new switch -n/--no-repeat which only downloads tweets once and exits
- OK, *now* fasttext links have been suppressed
- some old bugs crept back in. fixed

### v1.0
- improved subpage counter in the header
- reduced max tweet download limit to 200 as the twitter API doesnt let you do anymore
- added check to see if more than 99 pages will be generated
- add page_title_colour option to config.py
- happy with the documentation now!

### v0.9
- tweet usernames now go through character substition like tweets do
- useless hypertext links are now filtered and replaced with a placeholder
- blank lines now written to fill page, no more fasttext links inserted by TVs on line 24 by default if you don't specify any

### v0.8
- major feature addition: subpage support to handle a larger number of tweets
- added cycle_time parameter to config, this is cycle time between subpages
- added count argument to specify number of tweets to download
- a few more bug fixes: line 24 now writable and no more dropped tweets (they go to subpages instead)
- moved a few bits of code around for logic and readability

### v0.7
- compressed three different tweet output functions into one!
- fixed still display bug in write_home_timeline function

### v0.6
- New mode added: show latest tweets from a certain user
- cmd line switches changed to accommodate user mode
- logo looks less like a dog
- bug fix: tweet cleaning now works in search mode
- bug fix: added curly quote->normal quote replacement
- default config file wasn't formatted right

### v0.5.3
- Page number now user-customisable

### v0.5.2
- Fixed stupid bug where I'd used %S instead of %M for the time display. now they're correct

### v0.5.1
- more restructuring
- you can now customise the page title text

### v0.5
- extensive code restructuring, including splitting things out into modules
- code made a lot more "functional" resulting in saving some lines (it's a start anyway)

### v0.4
- emoji stripping added!
- implemented theme support: pick your own colour scheme!
- changed config file layout to help future improvements
- changing the folder structure to aid a more modular design
- added tti_path variable to config.py

### v0.3
- search mode added!
- all terminal output now sent correctly to stdout and stderr
- checking cmd line args for errors such as missing arguments or incompatible switched
- added some files to improve github community score

### v0.2
- improved error handling
- script now exits on twitter authentication error
- documentation much improved due to increaed public interest
- added argparse code with some options including quiet mode
- error checking added

### v0.1
- added some control code constants for readability and some more code comments
- graphical banner added, lots of colour, and more than one tweet on a screen
- writes a single tweet to a .TTI file now for output


teletext-twitter.py
# teletext-twitter - creates pages for vbit2 teletext system
# (c) Mark Pentler 2018 (https://github.com/mpentler)
# see README.md for details on getting it running or run with -h

import twitter
import time
import sys
import textwrap
import argparse
import re

# define our control codes here for easy use later
ESCAPE = chr(27)
RED = ESCAPE + chr(65)
CYAN = ESCAPE + chr(70)
DOUBLE_HEIGHT = ESCAPE + chr(77)
MOSAIC_CYAN = ESCAPE + chr(86)

# Read config.py for our Twitter access keys etc
config = {}
exec(open("config.py").read(), config)
twitter_object = twitter.Api(access_token_key = config["access_key"],
                      access_token_secret = config["access_secret"],
                      consumer_key = config["consumer_key"],
                      consumer_secret = config["consumer_secret"],
                      sleep_on_rate_limit = True) # so we don't hit the rate limit and raise an exception

def clean_line(line):
    emoji_pattern = re.compile("[" # our unicode ranges go here. this will need frequent tweaking
                              u"\U00002300-\U000023FF" # misc technical
                              u"\U000024C2-\U0001F251" # enclosed characters including flags
                              u"\U0001F300-\U0001F5FF" # symbols & pictographs
                              u"\U0001F600-\U0001F67F" # emoticons
                              u"\U0001F680-\U0001F9FF" # transport & map symbols
                               "]+", flags=re.UNICODE)
    line = emoji_pattern.sub(r'', line)

    line = line.replace("’", "'") # replacing some problematic characters here
    line = line.replace("_", "-") # the teletext English character set doesn't
    line = line.replace("#", "_") # support a lot of things!

    return line

def write_header(): # write a header for the page and pop a nice banner at the top
    with open(config["tti_path"] + "P153.tti", "w+") as file:
        file.write("DE,Autogenerated by Teletext-Twitter\r\n")
        file.write("PN,15300\r\n")
        file.write("SC,0000\r\n")
        file.write("PS,8000\r\n")
        file.write("OL,1," + DOUBLE_HEIGHT + "TELETEXT TWITTER" + MOSAIC_CYAN + (chr(127) * 22) + "\r\n")
        file.write("OL,3," + MOSAIC_CYAN + (chr(35) * 39) + "\r\n")

def write_timeline(): # grab the latest timeline - only 5 tweets for now
    statuses = twitter_object.GetHomeTimeline(count = 5)
    line_position = 4

    for status in statuses: # iterate through our responses
        tweet_time = time.strptime(status.created_at,"%a %b %d %H:%M:%S +0000 %Y")
        tweet_human_time = time.strftime("%d-%b-%Y %H:%S", tweet_time) # reformat time/date output
        tweet_username = status.user.screen_name
        tweet_text = textwrap.wrap(status.text, 39) # make sure our lines fit on the screen

        if (line_position + len(tweet_text) + 1) > 24: # are we going to fill the page?
            break # yep! dump the last tweet!

        with open(config["tti_path"] + "P153.tti", "a") as file:
            file.write("OL,{},".format(str(line_position)) + ("`" * (36-len(tweet_human_time)-len(tweet_username))) + "@{}".format(tweet_username) + " | " + "{}".format(tweet_human_time) + "\r\n")
            line_position += 1
            for line in tweet_text:
                line = clean_line(line)
                file.write("OL,{},".format(str(line_position)) + RED + "{}\r\n".format(line))
                line_position += 1

def write_search_term(search_term): # search recent tweets with a particular search term
    statuses = twitter_object.GetSearch(term=search_term, result_type="recent", count=5)
    line_position = 4

    for status in statuses: # iterate through our responses
        tweet_time = time.strptime(status.created_at,"%a %b %d %H:%M:%S +0000 %Y")
        tweet_human_time = time.strftime("%d-%b-%Y %H:%S", tweet_time) # reformat time/date output
        tweet_username = status.user.screen_name
        tweet_text = textwrap.wrap(status.text, 39) # make sure our lines fit on the screen

        if (line_position + len(tweet_text) + 1) > 24: # are we going to fill the page?
            break # yep! dump the last tweet!

        with open(config["tti_path"] + "P153.tti", "a") as file:
            file.write("OL,{},".format(str(line_position)) + ("`" * (36-len(tweet_human_time)-len(tweet_username))) + "@{}".format(tweet_username) + " | " + "{}".format(tweet_human_time) + "\r\n")
            line_position += 1
            for line in tweet_text:
                line = clean_line(line)
                file.write("OL,{},".format(str(line_position)) + RED + "{}\r\n".format(line))
                line_position += 1

def parse_args():
    parser = argparse.ArgumentParser(description="Reads your Twitter timeline and turns it into teletext pages for your Raspberry Pi.")

    parser.add_argument("-t", "--timeline", action="store_true", dest="home_timeline", help="download your latest home timeline", default=False)
    parser.add_argument("-s", "--search", action="store_true", dest="search", help="specify a term to search for")
    parser.add_argument("-q", "--query", type=str, help="a search query, hashtags supported if you put quoted around the string")
    parser.add_argument("-d", "--delay", type=int, default=60, help="seconds between timeline scrapes (minimum is 60 seconds - lower values have no effect)")
    parser.add_argument("-v", "--version", action="version", version="0.4-beta1")
    parser.add_argument("-Q", "--quiet", action="store_true", default=False, help="suppresses all output to the terminal except warnings and errors")

    args = parser.parse_args()
    args.delay = max(60, args.delay)

    if not args.home_timeline and not args.search:
        print("[!] No mode specified. Use -t or -s arguments. Exiting...", file=sys.stderr)
        sys.exit(1)

    if args.home_timeline and args.search:
        print("[!] Home timeline and search options cannot be selected together. Exiting...", file=sys.stderr)
        sys.exit(1)

    if args.search and not args.query:
        print("[!] Search option selected but no query specfied with -q. Exiting...", file=sys.stderr)
        sys.exit(1)

    return args

def main():
    args = parse_args()

    if not args.quiet:
        print("[*] teletext-twitter - (c) 2018 Mark Pentler (https://github.com/mpentler)", file=sys.stdout)

    while True:
        try:
            write_header()
            if args.home_timeline:
                if not args.quiet:
                    print("[*] Beginning home timeline scraping", file=sys.stdout)
                write_timeline()
            elif args.search:
                if not args.quiet:
                    print("[*] Getting recent tweets containing: " + args.query, file=sys.stdout)
                write_search_term(args.query)
            if not args.quiet:
                print("[*] Page updated. Waiting {} seconds until next scrape".format(args.delay), file=sys.stdout)
        except OSError as e:
            print("[!] Error accessing teletext data file, exiting: {}".format(e.strerror), file=sys.stderr)
            sys.exit(1)
        except twitter.error.TwitterError as e:
            for error in e.message:
                if error['code'] == 32:
                    print("[!] Authentication error accessing Twitter. Check config.py file and make sure your tokens are correct.", file=sys.stderr)
                    print("[!] Exiting...", file=sys.stderr)
                    sys.exit(1)
                print("[!] Error accessing your Twitter timeline: {}".format(error['message']), file=sys.stderr)
                print("[!] Trying again after specified delay", file=sys.stderr)
        time.sleep(args.delay)

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print("[*] Interrupted by user. Exiting...", file=sys.stdout)
        sys.exit(0)


teletext-twitter/output.py
# teletext-twitter - creates pages for vbit2 teletext system
# (c) Mark Pentler 2018 (https://github.com/mpentler)
# see README.md for details on getting it running or run with -h
# output module

from processor import *
import time

# define our control codes here for easy use later
ESCAPE = chr(27)
DOUBLE_HEIGHT = ESCAPE + chr(77)
SET_BACKGROUND = ESCAPE + chr(93)
TWITTER_BIRD = chr(41) + chr(108) + chr(39)
text_colours = {"red" : 65, "green" : 66, "yellow" : 67 , "blue" : 68, "magenta" : 69, "cyan" : 70, "white" : 71}
mosaic_colours = {"red" : 81, "green" : 82, "yellow" : 83, "blue" : 84, "magenta" : 85, "cyan" : 86, "white" : 87}

def write_tweet_info(file, line_num, username, timestamp, config):
    string = "OL," + str(line_num) + ","
    string += "`" * (35-len(timestamp)-len(username))
    string += ESCAPE + chr(text_colours[config["username_colour"]]) +"@" + username
    string += ESCAPE + chr(text_colours["white"]) + "|"
    string += ESCAPE + chr(text_colours[config["timestamp_colour"]]) + timestamp
    string += "\r\n"
    file.write(string)

def write_tweet_line(file, line_num, line, config):
    string = "OL," + str(line_num) + ","
    string += ESCAPE + chr(text_colours[config["tweet_colour"]]) + line
    string += "\r\n"
    file.write(string)

def write_header(file, subpage, max_subpages, config): # write a header for the page and pop a nice banner at the top
    page_title = config["page_title"] + " " + "{:02d}/{:02d}".format(subpage, max_subpages)
    logo_spacer = " " * (39 - (4 + len(page_title) + 5))
    if subpage == 1: # we only want these lines once
        file.write("DE,Autogenerated by Teletext-Twitter\r\n")
        file.write("PS,8000\r\n")
        file.write("CT," + str(config["cycle_time"]) + ",T\r\n")
    file.write("PN," + str(config["page_number"]) + "{:02}\r\n".format(subpage))
    file.write("SC,00" + "{:02}\r\n".format(subpage))
    file.write("OL,28,@@@|gpCu_@|wKpZA`UB_wLs_w}ww]_oh_wM@@\r\n") # define CLUT3:6 as twitter blue for enhanced logo
    file.write("OL,1," + ESCAPE + chr(text_colours[config["header_colour"]]) + SET_BACKGROUND +
               DOUBLE_HEIGHT + ESCAPE + chr(text_colours[config["page_title_colour"]]) + page_title +
               logo_spacer + ESCAPE + chr(mosaic_colours["cyan"]) + TWITTER_BIRD + "\r\n")
    file.write("OL,3," + ESCAPE + chr(mosaic_colours[config["header_separator"]]) + (chr(35) * 39) + "\r\n")

def write_enhancements(file, enhancements):
    for p in range(15): # up to 15 enhancement packets per subpage
        packet = "OL,26," + chr(0x40 + p) # first byte is designation code
        for e in range(13): # up to 13 triplets per enhancement packet
            if len(enhancements) > p*13+e:
                # combine parts of enhancement data into an 18 byte triplet, then slice it up into three 6 byte values to write to the row with bit 6 set
                triplet = enhancements[p*13+e][0] | ((enhancements[p*13+e][1]) << 6) | ((enhancements[p*13+e][2]) << 11)
                packet+=chr(0x40+(triplet&0x3F))+chr(0x40+((triplet>>6)&0x3F))+chr(0x40+((triplet>>12)&0x3F))
            elif e == 0:
                # TODO: generate packet of terminators if previous packet is full.
                return
            else:
                packet+=chr(0x7f)+chr(0x7f)+chr(0x7f)
        file.write(packet+"\r\n")

def write_tweets(twitter_object, mode, count, config, query=None): # grab the latest timeline
    if mode == "home":
        statuses = twitter_object.GetHomeTimeline(count=count)
    elif mode == "search":
        statuses = twitter_object.GetSearch(term=query, result_type="recent", count=count)
    elif mode == "user":
        statuses = twitter_object.GetUserTimeline(screen_name=query, count=count)

    filename = config["tti_path"] + "P" + str(config["page_number"]) + ".tti"
    subpage = 1

    # pre-calculate how many subpages we're going to
    # write - is there a better way of doing this?
    line_position = 4
    for status in statuses:
        tweet_text = tweet_remove_emojis(status.text)
        tweet_text = tweet_remove_urls(tweet_text)
        if mode == "search":
            tweet_text = tweet_highlight_query(tweet_text, query, config)
        tweet_text = charsub(tweet_text)
        tweet_text = textwrap.wrap(tweet_text, 38)
        post_length = len(tweet_text) + 1
        if (line_position + post_length) > 24:
            subpage += 1
            if subpage > 99:
                break # reached subpage limit - no point checking the rest
            line_position = 4
        line_position += post_length
    max_subpages = min(subpage, 99)

    # reset everything for the actual writing
    subpage = 1
    with open(filename, "w+", newline="") as file:
        write_header(file, subpage, max_subpages, config)
    line_position = 4
    subpage_enhancements = []

    for status in statuses: # iterate through our responses
        tweet_text = tweet_remove_emojis(status.text)
        tweet_text = tweet_remove_urls(tweet_text)
        if mode == "search":
            tweet_text = tweet_highlight_query(tweet_text, query, config)
        tweet_text = charsub(tweet_text)
        tweet_text = textwrap.wrap(tweet_text, 38) # make sure our lines fit on the screen
        tweet_time = time.strptime(status.created_at,"%a %b %d %H:%M:%S +0000 %Y")
        tweet_human_time = time.strftime("%d-%b-%Y %H:%S", tweet_time) # reformat time/date output
        tweet_username = charsub(status.user.screen_name)

        with open(filename, "a", newline="") as file:
            post_length = len(tweet_text) + 1 # how long is our next tweet? (including info line)
            if (line_position + post_length) > 24: # are we about to go over the page?
                file.write("FL,0,0,0,0,0,100\r\n")
                if subpage_enhancements:
                    write_enhancements(file, subpage_enhancements)
                subpage += 1 # start a new page
                subpage_enhancements = []
                if subpage > 99:
                    return # reached subpage limit - dump the rest
                write_header(file, subpage, max_subpages, config)
                line_position = 4 # and reset our cursor
            tweet_username_enhanced = charenhance(tweet_username,(37-len(tweet_human_time)-len(tweet_username)))
            if tweet_username_enhanced[1]:
                subpage_enhancements.append([line_position+40,4,0]) # active position to start of row
                subpage_enhancements += tweet_username_enhanced[1]
            write_tweet_info(file, line_position, tweet_username_enhanced[0], tweet_human_time, config)
            line_position += 1
            for line in tweet_text:
                tweet_text_line = charenhance(line,1)
                if tweet_text_line[1]:
                    subpage_enhancements.append([line_position+40,4,0]) # active position to start of row
                    subpage_enhancements += tweet_text_line[1]
                write_tweet_line(file, line_position, tweet_text_line[0], config)
                line_position += 1

    # finish the last subpage
    with open(filename, "a", newline="") as file:
        file.write("FL,0,0,0,0,0,100\r\n")
        if subpage_enhancements:
            write_enhancements(file, subpage_enhancements)


teletext-twitter/__main__.py
# teletext-twitter - creates pages for vbit2 teletext system
# (c) Mark Pentler 2018 (https://github.com/mpentler)
# see README.md for details on getting it running or run with -h
# main script

from output import *
import twitter
import time
import sys
import argparse

# Read config.py for our Twitter access keys etc
config = {}
exec(open("config.py").read(), config)
twitter_object = twitter.Api(access_token_key = config["access_key"],
                      access_token_secret = config["access_secret"],
                      consumer_key = config["consumer_key"],
                      consumer_secret = config["consumer_secret"],
                      sleep_on_rate_limit = True) # so we don't hit the rate limit and raise an exception

def parse_args():
    parser = argparse.ArgumentParser(description="Reads your Twitter timeline and turns it into teletext pages for your Raspberry Pi.")

    parser.add_argument("-m", "--mode", type=str, help="choose between different modes - home, user or search")
    parser.add_argument("-q", "--query", type=str, help="a search query, either a search term or a username. hashtags supported if you put quotes around the string")
    parser.add_argument("-c", "--count", type=int, default=5, help="number of tweets to download (default is 5, capped at 200)")
    parser.add_argument("-d", "--delay", type=int, default=60, help="seconds between timeline scrapes (default is 60 seconds - lower values have no effect)")
    parser.add_argument("-n", "--no-repeat", action="store_true", default=False, help="only download tweets once - overrules -d switch")
    parser.add_argument("-v", "--version", action="version", version="1.1.0")
    parser.add_argument("-Q", "--quiet", action="store_true", default=False, help="suppresses all output to the terminal except warnings and errors")

    args = parser.parse_args()
    args.count = min(args.count, 200)
    args.delay = max(60, args.delay)

    if args.mode == "search" and not args.query:
        print("[!] Search mode selected but no query specfied with -q. Exiting...", file=sys.stderr)
        sys.exit(1)
    if args.mode == "user" and not args.query:
        print("[!] User timeline mode selected but no username specified. Exiting...", file=sys.stderr)
        sys.exit(1)

    return args

def main():
    args = parse_args()

    if not args.quiet:
        print("[*] teletext-twitter - (c) 2018 Mark Pentler (https://github.com/mpentler)", file=sys.stdout)
        print("[*] additional support by Alistair Cree (https://zxnet.co.uk)", file=sys.stdout)

    while True:
        try:
            if args.mode == "home":
                if not args.quiet:
                    print("[*] Getting {} most recent tweets from your home timeline".format(args.count), file=sys.stdout)
                write_tweets(twitter_object, args.mode, args.count, config)
            elif args.mode == "search":
                if not args.quiet:
                    print("[*] Getting {} most recent tweets containing: {}".format(args.count, args.query), file=sys.stdout)
                write_tweets(twitter_object, args.mode, args.count, config, args.query)
            elif args.mode == "user":
                if not args.quiet:
                    print("[*] Getting {} most recent tweets from user: @{}".format(args.count, args.query), file=sys.stdout)
                write_tweets(twitter_object, args.mode, args.count, config, args.query)
            if not args.quiet:
                print("[*] Page updated. Waiting {} seconds until next scrape".format(args.delay), file=sys.stdout)
        except OSError as e:
            print("[!] Error accessing teletext data file, {}".format(e.strerror), file=sys.stderr)
            print("[!] Exiting...", file=sys.stderr)
            sys.exit(1)
        except twitter.error.TwitterError as e:
            for error in e.message:
                if error['code'] == 32:
                    print("[!] Authentication error accessing Twitter. Check config.py file and make sure your tokens are correct.", file=sys.stderr)
                    print("[!] Exiting...", file=sys.stderr)
                    sys.exit(1)
                print("[!] Error accessing your Twitter timeline: {}".format(error['message']), file=sys.stderr)
                print("[!] Trying again after specified delay", file=sys.stderr)
        if not args.no_repeat:
            time.sleep(args.delay)
        else:
            if not args.quiet:
                print("[*] No repeat mode enabled. Exiting...", file=sys.stdout)
            sys.exit(0)

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print("[*] Interrupted by user. Exiting...", file=sys.stdout)
        sys.exit(0)
